server/src/main/java/org/apache/cloudstack/diagnostics/DiagnosticsServiceImpl.java

```diff
@@ -126,10 +185,336 @@ protected String prepareShellCmd(String cmdType, String ipAddress, String option
+        }
+        return vmInstance;
+    }
+
+    private String getVMSshIp(final VMInstanceVO vmInstance) {
+        Map<String, String> accessDetails = networkManager.getSystemVMAccessDetails(vmInstance);
+        String controlIP = accessDetails.get(NetworkElementCommand.ROUTER_IP);
+        if (StringUtils.isBlank(controlIP)) {
+            throw new CloudRuntimeException("Unable to find system vm ssh/control IP for  vm with ID: " + vmInstance.getId());
+        }
+        return controlIP;
+    }
+
+    @Override
+    public boolean start() {
+        super.start();
+        return true;
+    }
+
+    @Override
+    public boolean configure(final String name, final Map<String, Object> params) throws ConfigurationException {
+        if (EnableGarbageCollector.value()) {
+            backgroundPollManager.submitTask(new GCBackgroundTask(this));
+        }
+        return true;
+    }
+
+    public static final class GCBackgroundTask extends ManagedContextRunnable implements BackgroundPollTask {
+        private DiagnosticsServiceImpl serviceImpl;
+
+        public GCBackgroundTask(DiagnosticsServiceImpl serviceImpl) {
+            this.serviceImpl = serviceImpl;
+        }
+
+        private static void deleteOldDiagnosticsFiles(File directory, String storeName) {
+            final File[] fileList = directory.listFiles();
+            if (fileList != null) {
+                String msg = String.format("Found %s diagnostics files in store %s for garbage collection", fileList.length, storeName);
+                LOGGER.info(msg);
+                for (File file : fileList) {
+                    if (file.isFile()) {
+                        if (MaximumFileAgeforGarbageCollection.value() <= getTimeDifference(file)) {
+                            boolean success = file.delete();
+                            LOGGER.info(file.getName() + " delete status: " + success);
+                        }
+                    }
+                }
+            }
+        }
+
+        @Override
+        protected void runInContext() {
+            List<DataCenterVO> dcList = serviceImpl.dataCenterDao.listEnabledZones();
+            for (DataCenterVO vo: dcList) {
+                // Get All Image Stores in current running Zone
+                List<DataStore> storeList = serviceImpl.storeMgr.getImageStoresByScope(new ZoneScope(vo.getId()));
+                for (DataStore store : storeList) {
+                    String mountPoint = null;
+                    try {
+                        mountPoint = serviceImpl.mountManager.getMountPoint(store.getUri(), null);
+                        if (StringUtils.isNotBlank(mountPoint)) {
+                            File directory = new File(mountPoint + "/" + DIAGNOSTICS_DIRECTORY);
+                            if (directory.isDirectory()) {
+                                deleteOldDiagnosticsFiles(directory, store.getName());
+                            }
+                        }
```

Comment:
rather high level of nesting/cyclic complexity. Can you extract this bit, that would leave
```
try {
    something();
} finally {
    do_something();
}